src/Razor/Microsoft.AspNetCore.Razor.Language/src/Components/ComponentDesignTimeNodeWriter.cs

```diff
@@ -418,22 +418,49 @@ public override void WriteComponent(CodeRenderingContext context, ComponentInter
             }
             else
             {
+                var parameters = GetTypeInferenceMethodParameters(node.TypeInferenceNode);
+
+                // If this component is going to cascade any of its generic types, we have to split its type inference
+                // into two parts. First we call an inference method that captures all the parameters in local variables,
+                // then we use those to call the real type inference method that emits the component. The reason for this
+                // is so the captured variables can be used by descendants without re-evaluating the expressions.
+                CodeWriterExtensions.CSharpCodeWritingScope? typeInferenceCaptureScope = null;
+                if (node.Component.SuppliesCascadingGenericParameters())
+                {
+                    typeInferenceCaptureScope = context.CodeWriter.BuildScope();
+                    context.CodeWriter.Write(node.TypeInferenceNode.FullTypeName);
+                    context.CodeWriter.Write(".");
+                    context.CodeWriter.Write(node.TypeInferenceNode.MethodName);
+                    context.CodeWriter.Write("_CaptureParameters(");
+                    var isFirst = true;
+                    foreach (var parameter in parameters.Where(p => p.UsedForTypeInference))
+                    {
+                        if (isFirst)
+                        {
+                            isFirst = false;
+                        }
+                        else
+                        {
+                            context.CodeWriter.Write(", ");
+                        }
+
+                        WriteTypeInferenceMethodParameterInnards(context, parameter);
+                        context.CodeWriter.Write(", out var ");
+
+                        var variableName = $"__typeInferenceArg_{_scopeStack.Depth}_{parameter.ParameterName}";
```

Comment:
Can you help me understand what role including the `scopeStack.Depth` property here plays?

Reply:
I suspect this is because you can have multiple nested components that perform type inference within the same file and you need to be able to distinguish between them.

Reply:
The code we generated includes nested lambdas for each level of child content. It's possible that an ancestor or descendant also tries to do type inference for a parameter with the same name. We can only be confident that parameters are uniquely named within their own level.

Example - a "treeview" component might have a `Node<T>` component with some parameter `T Item`, like:

```razor
<Node Item="rootItem">
    <Node Item="childItem"> ... other content here ... </Node>
</Node>
```

If `Node<T>` declares that it cascades the `T` parameter, we'd generate code like:

```cs
EmitNodeComponent_CaptureParameters(rootItem, out var __typeInferenceArg_0_Item);
EmitNodeComponent(builder0, __typeInferenceArg_0_Item, builder1 =>
{
    EmitNodeComponent_CaptureParameters(childItem, out var __typeInferenceArg_1_Item);
    EmitNodeComponent(builder1, __typeInferenceArg_1_Item, builder2 =>
    {
        ... other content here ...
    });
});
```

The stack depth is enough to disambiguate the different variables. In this specific example it wouldn't really matter if a child variable shadowed an ancestor variable, but in more general cases an arbitrary descendant might be using an arbitrary ancestor for type inference, so we need to keep them distinct.

Note that we don't have to disambiguate siblings, e.g.:

```
<Node Item="sibling0">... content 0 ...</Node>
<Node Item="sibling1">... content 1 ...</Node>
```

... because the code we generate looks like the following, using sibling-level C# scopes to allow multiple variables with the same name:

```cs
{
    EmitNodeComponent_CaptureParameters(sibling0, out var __typeInferenceArg_0_Item);
    EmitNodeComponent(builder0, __typeInferenceArg_0_Item, builder1 => { ... content 1 ... });
}
{
    EmitNodeComponent_CaptureParameters(sibling0, out var __typeInferenceArg_0_Item);
    EmitNodeComponent(builder0, __typeInferenceArg_0_Item, builder1 => { ... content 1 ... });
}
```